// livekit-server-sdk/dist/SipClient.js
import { Duration } from "@bufbuild/protobuf";
import {
  CreateSIPDispatchRuleRequest,
  CreateSIPInboundTrunkRequest,
  CreateSIPOutboundTrunkRequest,
  CreateSIPParticipantRequest,
  CreateSIPTrunkRequest,
  DeleteSIPDispatchRuleRequest,
  DeleteSIPTrunkRequest,
  ListSIPDispatchRuleRequest,
  ListSIPDispatchRuleResponse,
  ListSIPInboundTrunkRequest,
  ListSIPInboundTrunkResponse,
  ListSIPOutboundTrunkRequest,
  ListSIPOutboundTrunkResponse,
  ListSIPTrunkRequest,
  ListSIPTrunkResponse,
  SIPDispatchRule,
  SIPDispatchRuleDirect,
  SIPDispatchRuleIndividual,
  SIPDispatchRuleInfo,
  SIPInboundTrunkInfo,
  SIPOutboundTrunkInfo,
  SIPParticipantInfo,
  SIPTransport,
  SIPTrunkInfo,
  TransferSIPParticipantRequest
} from "@livekit/protocol";
import ServiceBase from "./ServiceBase.js";
import { TwirpRpc, livekitPackage } from "./TwirpRPC.js";
const svc = "SIP";
class SipClient extends ServiceBase {
  /**
   * @param host - hostname including protocol. i.e. 'https://<project>.livekit.cloud'
   * @param apiKey - API Key, can be set in env var LIVEKIT_API_KEY
   * @param secret - API Secret, can be set in env var LIVEKIT_API_SECRET
   */
  constructor(host, apiKey, secret) {
    super(apiKey, secret);
    this.rpc = new TwirpRpc(host, livekitPackage);
  }
  /**
   * @param number - phone number of the trunk
   * @param opts - CreateSipTrunkOptions
   * @deprecated use `createSipInboundTrunk` or `createSipOutboundTrunk`
   */
  async createSipTrunk(number, opts) {
    let inboundAddresses;
    let inboundNumbers;
    let inboundUsername = "";
    let inboundPassword = "";
    let outboundAddress = "";
    let outboundUsername = "";
    let outboundPassword = "";
    let name = "";
    let metadata = "";
    if (opts !== void 0) {
      inboundAddresses = opts.inbound_addresses;
      inboundNumbers = opts.inbound_numbers;
      inboundUsername = opts.inbound_username || "";
      inboundPassword = opts.inbound_password || "";
      outboundAddress = opts.outbound_address || "";
      outboundUsername = opts.outbound_username || "";
      outboundPassword = opts.outbound_password || "";
      name = opts.name || "";
      metadata = opts.metadata || "";
    }
    const req = new CreateSIPTrunkRequest({
      name,
      metadata,
      inboundAddresses,
      inboundNumbers,
      inboundUsername,
      inboundPassword,
      outboundNumber: number,
      outboundAddress,
      outboundUsername,
      outboundPassword
    }).toJson();
    const data = await this.rpc.request(
      svc,
      "CreateSIPTrunk",
      req,
      await this.authHeader({}, { admin: true })
    );
    return SIPTrunkInfo.fromJson(data, { ignoreUnknownFields: true });
  }
  /**
   * @param name - human-readable name of the trunk
   * @param numbers - phone numbers of the trunk
   * @param opts - CreateSipTrunkOptions
   */
  async createSipInboundTrunk(name, numbers, opts) {
    let allowedAddresses;
    let allowedNumbers;
    let authUsername = "";
    let authPassword = "";
    let metadata = "";
    let headers = {};
    let headersToAttributes = {};
    if (opts !== void 0) {
      allowedAddresses = opts.allowed_addresses;
      allowedNumbers = opts.allowed_numbers;
      authUsername = opts.auth_username || "";
      authPassword = opts.auth_password || "";
      metadata = opts.metadata || "";
      headers = opts.headers || {};
      headersToAttributes = opts.headersToAttributes || {};
    }
    const req = new CreateSIPInboundTrunkRequest({
      trunk: new SIPInboundTrunkInfo({
        name,
        numbers,
        metadata,
        allowedAddresses,
        allowedNumbers,
        authUsername,
        authPassword,
        headers,
        headersToAttributes
      })
    }).toJson();
    const data = await this.rpc.request(
      svc,
      "CreateSIPInboundTrunk",
      req,
      await this.authHeader({}, { admin: true })
    );
    return SIPInboundTrunkInfo.fromJson(data, { ignoreUnknownFields: true });
  }
  /**
   * @param name - human-readable name of the trunk
   * @param address - hostname and port of the SIP server to dial
   * @param numbers - phone numbers of the trunk
   * @param opts - CreateSipTrunkOptions
   */
  async createSipOutboundTrunk(name, address, numbers, opts) {
    let authUsername = "";
    let authPassword = "";
    let transport = SIPTransport.SIP_TRANSPORT_AUTO;
    let metadata = "";
    let headers = {};
    let headersToAttributes = {};
    if (opts !== void 0) {
      authUsername = opts.auth_username || "";
      authPassword = opts.auth_password || "";
      transport = opts.transport || SIPTransport.SIP_TRANSPORT_AUTO;
      metadata = opts.metadata || "";
      headers = opts.headers || {};
      headersToAttributes = opts.headersToAttributes || {};
    }
    const req = new CreateSIPOutboundTrunkRequest({
      trunk: new SIPOutboundTrunkInfo({
        name,
        address,
        numbers,
        metadata,
        transport,
        authUsername,
        authPassword,
        headers,
        headersToAttributes
      })
    }).toJson();
    const data = await this.rpc.request(
      svc,
      "CreateSIPOutboundTrunk",
      req,
      await this.authHeader({}, { admin: true })
    );
    return SIPOutboundTrunkInfo.fromJson(data, { ignoreUnknownFields: true });
  }
  /**
   * @deprecated use `listSipInboundTrunk` or `listSipOutboundTrunk`
   */
  async listSipTrunk() {
    const req = {};
    const data = await this.rpc.request(
      svc,
      "ListSIPTrunk",
      new ListSIPTrunkRequest(req).toJson(),
      await this.authHeader({}, { admin: true })
    );
    return ListSIPTrunkResponse.fromJson(data, { ignoreUnknownFields: true }).items ?? [];
  }
  async listSipInboundTrunk() {
    const req = {};
    const data = await this.rpc.request(
      svc,
      "ListSIPInboundTrunk",
      new ListSIPInboundTrunkRequest(req).toJson(),
      await this.authHeader({}, { admin: true })
    );
    return ListSIPInboundTrunkResponse.fromJson(data, { ignoreUnknownFields: true }).items ?? [];
  }
  async listSipOutboundTrunk() {
    const req = {};
    const data = await this.rpc.request(
      svc,
      "ListSIPOutboundTrunk",
      new ListSIPOutboundTrunkRequest(req).toJson(),
      await this.authHeader({}, { admin: true })
    );
    return ListSIPOutboundTrunkResponse.fromJson(data, { ignoreUnknownFields: true }).items ?? [];
  }
  /**
   * @param sipTrunkId - sip trunk to delete
   */
  async deleteSipTrunk(sipTrunkId) {
    const data = await this.rpc.request(
      svc,
      "DeleteSIPTrunk",
      new DeleteSIPTrunkRequest({ sipTrunkId }).toJson(),
      await this.authHeader({}, { admin: true })
    );
    return SIPTrunkInfo.fromJson(data, { ignoreUnknownFields: true });
  }
  /**
   * @param rule - sip dispatch rule
   * @param opts - CreateSipDispatchRuleOptions
   */
  async createSipDispatchRule(rule, opts) {
    let trunkIds;
    let hidePhoneNumber = false;
    let name = "";
    let metadata = "";
    let ruleProto = void 0;
    if (opts !== void 0) {
      trunkIds = opts.trunkIds;
      hidePhoneNumber = opts.hidePhoneNumber || false;
      name = opts.name || "";
      metadata = opts.metadata || "";
    }
    if (rule.type == "direct") {
      ruleProto = new SIPDispatchRule({
        rule: {
          case: "dispatchRuleDirect",
          value: new SIPDispatchRuleDirect({
            roomName: rule.roomName,
            pin: rule.pin || ""
          })
        }
      });
    } else if (rule.type == "individual") {
      ruleProto = new SIPDispatchRule({
        rule: {
          case: "dispatchRuleIndividual",
          value: new SIPDispatchRuleIndividual({
            roomPrefix: rule.roomPrefix,
            pin: rule.pin || ""
          })
        }
      });
    }
    const req = new CreateSIPDispatchRuleRequest({
      rule: ruleProto,
      trunkIds,
      hidePhoneNumber,
      name,
      metadata
    }).toJson();
    const data = await this.rpc.request(
      svc,
      "CreateSIPDispatchRule",
      req,
      await this.authHeader({}, { admin: true })
    );
    return SIPDispatchRuleInfo.fromJson(data, { ignoreUnknownFields: true });
  }
  async listSipDispatchRule() {
    const req = {};
    const data = await this.rpc.request(
      svc,
      "ListSIPDispatchRule",
      new ListSIPDispatchRuleRequest(req).toJson(),
      await this.authHeader({}, { admin: true })
    );
    return ListSIPDispatchRuleResponse.fromJson(data, { ignoreUnknownFields: true }).items ?? [];
  }
  /**
   * @param sipDispatchRuleId - sip trunk to delete
   */
  async deleteSipDispatchRule(sipDispatchRuleId) {
    const data = await this.rpc.request(
      svc,
      "DeleteSIPDispatchRule",
      new DeleteSIPDispatchRuleRequest({ sipDispatchRuleId }).toJson(),
      await this.authHeader({}, { admin: true })
    );
    return SIPDispatchRuleInfo.fromJson(data, { ignoreUnknownFields: true });
  }
  /**
   * @param sipTrunkId - sip trunk to use for the call
   * @param number - number to dial
   * @param roomName - room to attach the call to
   * @param opts - CreateSipParticipantOptions
   */
  async createSipParticipant(sipTrunkId, number, roomName, opts) {
    let participantIdentity = "";
    let participantName = "";
    let participantMetadata = "";
    let dtmf = "";
    let playRingtone = false;
    let playDialtone = false;
    let hidePhoneNumber = false;
    let ringingTimeout = void 0;
    let maxCallDuration = void 0;
    let enableKrisp = void 0;
    if (opts !== void 0) {
      participantIdentity = opts.participantIdentity || "";
      participantName = opts.participantName || "";
      participantMetadata = opts.participantMetadata || "";
      dtmf = opts.dtmf || "";
      playRingtone = opts.playRingtone || false;
      playDialtone = opts.playDialtone || playRingtone;
      hidePhoneNumber = opts.hidePhoneNumber || false;
      ringingTimeout = opts.ringingTimeout || void 0;
      maxCallDuration = opts.maxCallDuration || void 0;
      enableKrisp = opts.enableKrisp || void 0;
    }
    const req = new CreateSIPParticipantRequest({
      sipTrunkId,
      sipCallTo: number,
      roomName,
      participantIdentity,
      participantName,
      participantMetadata,
      dtmf,
      playRingtone: playDialtone,
      playDialtone,
      hidePhoneNumber,
      ringingTimeout: ringingTimeout ? new Duration({ seconds: BigInt(ringingTimeout) }) : void 0,
      maxCallDuration: maxCallDuration ? new Duration({ seconds: BigInt(maxCallDuration) }) : void 0,
      enableKrisp
    }).toJson();
    const data = await this.rpc.request(
      svc,
      "CreateSIPParticipant",
      req,
      await this.authHeader({}, { call: true })
    );
    return SIPParticipantInfo.fromJson(data, { ignoreUnknownFields: true });
  }
  /**
   * @param roomName - room the SIP participant to transfer is connectd to
   * @param participantIdentity - identity of the SIP participant to transfer
   * @param transferTo - SIP URL to transfer the participant to
   */
  async transferSipParticipant(roomName, participantIdentity, transferTo, opts) {
    let playDialtone = false;
    if (opts !== void 0) {
      playDialtone = opts.playDialtone || false;
    }
    const req = new TransferSIPParticipantRequest({
      participantIdentity,
      roomName,
      transferTo,
      playDialtone
    }).toJson();
    await this.rpc.request(
      svc,
      "TransferSIPParticipant",
      req,
      await this.authHeader({ roomAdmin: true, room: roomName }, { call: true })
    );
  }
}
export {
  SipClient
};
